src/main/java/hudson/plugins/clover/CloverPublisher.java

```diff
@@ -175,14 +199,105 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListen
         return true;
     }
 
+    @Override
+    public void perform(Run<?, ?> run, FilePath workspace, Launcher launcher, TaskListener listener) throws
+            InterruptedException, IOException {
+
+        final EnvVars env = run.getEnvironment(listener);
+        final File buildRootDir = run.getRootDir(); // should this top level?
+        final FilePath buildTarget = new FilePath(buildRootDir);
+        String reportDir = env.expand(cloverReportDir);
+        FilePath coverageReportDir = workspace.child(reportDir);
+        try {
+            listener.getLogger().println("Publishing Clover coverage report...");
+
+            // search one deep for the report dir, if it doesn't exist.
+            if (!coverageReportDir.exists()) {
+                coverageReportDir = findOneDirDeep(workspace, reportDir);
+            }
+
+            // if the build has failed, then there's not
+            // much point in reporting an error
+            final boolean buildFailure = run.getResult().isWorseOrEqualTo(Result.FAILURE);
+            final boolean missingReport = !coverageReportDir.exists();
+
+            if (buildFailure && missingReport) {
+                listener.getLogger().println("No Clover report will be published due to a " + (buildFailure ? "Build Failure" : "missing report"));
+            }
+
+            final boolean htmlExists = copyHtmlReport(coverageReportDir, buildTarget, listener);
+            final boolean xmlExists = copyXmlReport(coverageReportDir, buildTarget, listener, env.expand(getCloverReportFileName()));
+
+            if (htmlExists) {
+                // only add the HTML build action, if the HTML report is available
+                run.addAction(new CloverHtmlBuildAction(buildTarget));
+            }
+            processCloverXml(run, workspace, listener, coverageReportDir, buildTarget);
+
+        } catch (IOException e) {
+            Util.displayIOException(e, listener);
+            e.printStackTrace(listener.fatalError("Unable to copy coverage from " + coverageReportDir + " to " + buildTarget));
+            run.setResult(Result.FAILURE);
+        }
+    }
+
     /**
      * Process the clover.xml from the build directory. The clover.xml must have been already copied to the build dir.
      *
+     * @deprecated see {@link #processCloverXml(Run, FilePath, TaskListener, FilePath, FilePath)}
      */
+    @Deprecated
```

Comment:
It's a weird thing to deprecate a method and add new code on it. Perhaps I'm missing something.